### PR TITLE
add pax tests

### PR DESCRIPTION
**Pull request by khshah6**
## What
Adding two single host pax nightly tests

## Oneshot test results
[LmSpmd2B](https://pantheon.corp.google.com/kubernetes/job/us-central2/xl-ml-test-us-central2/default/pax-lmspmd2b-func-v4-8-1vm-vdtmp/logs?mods=allow_workbench_image_override&project=xl-ml-test)

### CODE REVIEW

File: tests/pax/common.libsonnet [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        gsutil cp gs://pax-on-cloud-tpu-project/wheels/20220814/paxml-nightly+20220814-py3-none-any.whl .
+        pip install paxml-nightly+20220814-py3-none-any.whl
+        pip install praxis
+      |||,
+    },    
+
+
+    metricConfig+: {
+      sourceMap+:: {
+        tensorboard+: {
```
Comment (will-cromar) [inline tests/pax/common.libsonnet:43]: This assertion shouldn't be necessary -- it only applies to TF with Keras.

File: tests/pax/common.libsonnet [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+local common = import '../common.libsonnet';
+local experimental = import '../experimental.libsonnet';
+local mixins = import 'templates/mixins.libsonnet';
+local tpus = import 'templates/tpus.libsonnet';
+
+{
+  PaxTest:: common.CloudAcceleratorTest + experimental.BaseTpuVmMixin {
```
Comment (will-cromar) [inline tests/pax/common.libsonnet:21]: I know it's clunky, but you should include some sort of base command that cleans up the TPU after the test script runs (`/scripts/cleanup.sh`). JAX is probably the closest example:

https://github.com/GoogleCloudPlatform/ml-testing-accelerators/blob/20c22a3517804809ba60c48354043937a11447b7/tests/jax/common.libsonnet#L64-L85

**Comment by khshah6** [comment]
I'm not sure why, but I'm hitting some strange Jax error when running the oneshot test:

```
/usr/local/lib/python3.7/dist-packages/jax/__init__.py:27: UserWarning: cloud_tpu_init failed: KeyError('<!DOCTYPE html>\n<html lang=en>\n <meta charset=utf-8>\n <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">\n <title>Error 404 (Not Found)!!1</title>\n <style>\n *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}\n </style>\n <a href=//www.google.com/><span id=logo aria-label=Google></span></a>\n <p><b>404.</b> <ins>That’s an error.</ins>\n <p>The requested URL <code>/computeMetadata/v1/instance/attributes/accelerator-type</code> was not found on this server. <ins>That’s all we know.</ins>\n')
```

I'm able to install the same packages and don't get this error when running the model manually.

Full oneshot test logs [here](https://pantheon.corp.google.com/kubernetes/job/us-central2/xl-ml-test-us-central2/default/pax-lmspmd2b-func-v4-8-1vm-j2qg6/logs?mods=allow_workbench_image_override&project=xl-ml-test)

**Comment by khshah6** [comment]
For some reason, the oneshot test logs are showing `InitializeTpuLibrary` errors, while running the same installation steps manually succeeds.

Errors:
```
Failed to get accelerator type with error: Failed to fetch URL (http status: 404) No errorI0000 00:00:1661883622.246389 525 common_lib.cc:145] Failed to fetch URL on try 1 out of 6: Failed to fetch URL (http status: 404) No error
```
```
Failed to get worker list with error: Failed to fetch URL (http status: 404) No errorF0830 18:20:25.258814 525 enforce_kernel_ipv6_support.cc:72] Check failed: loopback6_ok || !loopback4_ok This machine (pax-lmspmd2b-func-v4-8-1vm-trqkt-vpwvs) does not have an IPv6 loopback (::1) interface. This configuration is obsolete. You may temporarily set --enforce_kernel_ipv6_support=false to bypass this check, but code elsewhere in g3 may fail as result. See gl_________________30/topics/localhost for details.
```

Oneshot test logs [here](https://pantheon.corp.google.com/kubernetes/job/us-central2/xl-ml-test-us-central2/default/pax-lmspmd2b-func-v4-8-1vm-trqkt/logs?mods=allow_workbench_image_override&project=xl-ml-test)
For comparison, manual test logs [here](https://paste.googleplex.com/6584232448098304)

**Review by aireenmei — APPROVED** [review]

File: tests/pax/nightly/common.libsonnet [hunk 3]
```diff
@@ -0,0 +1,61 @@
+local common = import '../common.libsonnet';
+
+{
+  NightlyPaxTest:: common.PaxTest {
+    local config = self,
+    expPath:: '',
+    extraFlags:: '',
```
Comment (will-cromar) [inline tests/pax/nightly/common.libsonnet:7]: This should ideally be a list. It'll be easier to append flags as needed without having to worry about spacing. You can combine the list of flags into a string when you substitute it with something like `std.join(" ", config.extraFlags)`

File: tests/pax/nightly/common.libsonnet [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+      if [ -f praxis*.whl -a -f paxml*.whl ]; then
+          echo "Nightly builds succeeded."
+      else 
+          echo "Nighlty builds failed or are pending."
+          exit 1
+      fi 
+
+      # install pax and dependencies
+      pip install -U pip
```
Comment (will-cromar) [inline tests/pax/nightly/common.libsonnet:33]: I don't recommend updating pip with pip. Do you need a more recent version of `pip` than is on the base image?

File: tests/pax/nightly/common.libsonnet [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      pip install praxis*.whl
+      pip install paxml*.whl
+      sudo pip uninstall --yes jax jaxlib libtpu-nightly libtpu libtpu_tpuv4 tensorflow
```
Comment (will-cromar) [inline tests/pax/nightly/common.libsonnet:36]: Can you install praxis and paxml with `--no-deps` instead of uninstalling these dependencies? The `tpu-vm-base` image does not install any of these.

File: tests/pax/nightly/common.libsonnet [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      git clone https://github.com/google/jax.git
+      cd jax/
+      pip install .
+      cd ..
+      pip install --pre -U jaxlib -f https://storage.googleapis.com/jax-releases/jaxlib_nightly_releases.html
+      pip install -U libtpu-nightly -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
```
Comment (will-cromar) [inline tests/pax/nightly/common.libsonnet:42]: Does the nightly `jaxlib` target the nightly `libtpu`? It looks like you overwrite `libtpu.so` below. Can you just install the pinned libtpu version with `pip`?

**Review by will-cromar — CHANGES_REQUESTED** [review]

File: tests/pax/nightly/common.libsonnet (hunk 5 above)
Comment (khshah6) [reply tests/pax/nightly/common.libsonnet:36]: sure

File: tests/pax/nightly/common.libsonnet (hunk 6 above)
Comment (A9isha) [reply tests/pax/nightly/common.libsonnet:42]: yes the `jaxlib` has dependency on `libtpu`. At the moment, the available `libtpu` latest version is outdated, they are working on releasing the latest `libtpu`, so for now we are using a manually built version of `libtpu`. That's why I think we should clarify that this is temporary and we don't need both in the future

File: tests/pax/nightly/common.libsonnet (hunk 5 above)
Comment (khshah6) [reply tests/pax/nightly/common.libsonnet:36]: the only issue is that there are a few other dependencies we would have to manually install

for pax: [
        'praxis', 'pyglove', 'absl-py', 'jax', 'tensorflow==2.7.3',
        'tensorflow-text==2.7.3', 'numpy', 'flax',
        'clu', 'lingvo', 't5', 'seqio-nightly', 'tensorstore',
        'etils', 'orbax'
    ]

for praxis: [
        'protobuf', 'absl-py', 'fiddle-config', 'numpy', 'tensorflow', 'lingvo',
        'flax', 'jax', 'optax', 'optax-shampoo', 'jax-bitempered-loss',
        'einops', 't5x', 'clu'
   ]

File: tests/pax/nightly/common.libsonnet (hunk 4 above)
Comment (khshah6) [reply tests/pax/nightly/common.libsonnet:33]: I don't think so

File: tests/pax/nightly/common.libsonnet (hunk 6 above)
Comment (khshah6) [reply tests/pax/nightly/common.libsonnet:42]: yeah good catch, we want latest libtpu.so, which is built at head at gs://pax-on-cloud-tpu-project/libtpu/libtpu.so

File: tests/pax/nightly/lmtransformeradam.libsonnet [hunk 7]
```diff
@@ -0,0 +1,18 @@
+local common = import 'common.libsonnet';
+local mixins = import 'templates/mixins.libsonnet';
+local tpus = import 'templates/tpus.libsonnet';
+local utils = import 'templates/utils.libsonnet';
```
[the remark this answers is not in the document]
Comment (khshah6) [reply tests/pax/nightly/lmtransformeradam.libsonnet:4]: ah good call I forgot to remove the import

File: tests/pax/nightly/common.libsonnet (hunk 5 above)
Comment (khshah6) [reply tests/pax/nightly/common.libsonnet:39]: @will-cromar I addressed all the feedback except this in the last commit. Should I add a step to `pip install` these dependencies `pyglove, absl-py, tensorflow, tensorflow-text, numpy, flax, clu, lingvo, t4, sequio-nightly, tensorstore, etlist, orbax, fiddle-config, optax, optax-shampoo, jax-bitempered-loss, eniops, t5x` and add the `no-deps` flags, or keep it as is?

File: tests/pax/nightly/common.libsonnet [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      pip install -U pip
+      pip install praxis*.whl
+      pip install paxml*.whl
+      sudo pip uninstall jax jaxlib libtpu-nightly libtpu libtpu_tpuv4 tensorflow  -y
+      git clone https://github.com/google/jax.git
+      cd jax/
+      pip install .
+      cd ..
+      pip install --pre -U jaxlib -f https://storage.googleapis.com/jax-releases/jaxlib_nightly_releases.html
+      pip install -U libtpu-nightly -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
```
[the remark this answers is not in the document]
Comment (khshah6) [reply tests/pax/nightly/common.libsonnet:46]: the latest nightly libtpu is good now!

File: tests/pax/nightly/common.libsonnet (hunk 6 above)
Comment (khshah6) [reply tests/pax/nightly/common.libsonnet:46]: actually latest nightly libtpu worked, I modified the script to install that and not overwrite it

File: tests/pax/nightly/common.libsonnet (hunk 5 above)
Comment (will-cromar) [reply tests/pax/nightly/common.libsonnet]: Discussed offline. Removing some redundant uninstalls and leaving this as a short-term fix.
